### PR TITLE
Disable unwinding for emscripten again

### LINKED ISSUE #85821: Emscripten wasm32 compilation broken
I tried this code:

```rust
fn main() {
    println!("Hello, world!");
}
```

I expected to see this happen: Compiling with `rustc --target=wasm32-unknown-emscripten main.rs` , after the emsdk is activated and emcc works, to compile properly

Instead, this happened: Prints this error:
```
  = note: error: undefined symbol: __gxx_personality_v0 (referenced by top-level compiled C/C++ code)
          warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
          warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
          warning: ___gxx_personality_v0 may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
          Error: Aborting compilation due to previous errors
          emcc: error: 'C:/Users/camer/Documents/Tech/emsdk/node/14.15.5_64bit/bin/node.exe C:\Users\camer\Documents\Tech\emsdk\upstream\emscripten\src\compiler.js C:\Users\camer\AppData\Local\Temp\tmprxn1kc58.txt' failed (returned 1)
```

I've tried adding `-lstdc++` to no avail

### Meta

`rustc --version --verbose`:
```
rustc 1.54.0-nightly (f58631b45 2021-05-28)
binary: rustc
commit-hash: f58631b4503b5cd34163cf9c3ff19dc3e0c8a09e
commit-date: 2021-05-28
host: x86_64-pc-windows-msvc
release: 1.54.0-nightly
LLVM version: 12.0.1
```

<details><summary>Backtrace</summary>
<p>

```
error: linking with `emcc.bat` failed: exit code: 1
  |
  = note: "cmd" "/c" "emcc.bat" "-s" "DISABLE_EXCEPTION_CATCHING=0" "-L" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib" "-L" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\self-contained" "main.main.7rcbfp3g-cgu.0.rcgu.o" "main.main.7rcbfp3g-cgu.1.rcgu.o" "main.main.7rcbfp3g-cgu.2.rcgu.o" "main.main.7rcbfp3g-cgu.3.rcgu.o" "main.main.7rcbfp3g-cgu.4.rcgu.o" "main.main.7rcbfp3g-cgu.5.rcgu.o" "main.main.7rcbfp3g-cgu.6.rcgu.o" "main.main.7rcbfp3g-cgu.7.rcgu.o" "-o" "main.js" "-s" "EXPORTED_FUNCTIONS=[\"_main\",\"_rust_eh_personality\"]" "main.4s37gsrti678ik8u.rcgu.o" "-O0" "--memory-init-file" "0" "-g0" "-s" "DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=[]" "-L" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libstd-fb3d38fb81b27f3a.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libpanic_unwind-d2cae8b45410ca8e.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libstd_detect-59eadd895a0069e7.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\librustc_demangle-034447b1a63c6379.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libhashbrown-3b180b71fa933791.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\librustc_std_workspace_alloc-5d053d604e1da3fd.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libunwind-835bbc670368e525.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libcfg_if-8111a663608e727d.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\liblibc-35dde82b6021f4d8.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\liballoc-a187fd59d7a86f05.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\librustc_std_workspace_core-81a0f0dbfa7b5ebf.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libcore-ced9603eedaff686.rlib" "C:\\Users\\camer\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\wasm32-unknown-emscripten\\lib\\libcompiler_builtins-2960d62a9b1d31c1.rlib" "-l" "c" "-s" "ERROR_ON_UNDEFINED_SYMBOLS=1" "-s" "ASSERTIONS=1" "-s" "ABORTING_MALLOC=0" "-Wl,--fatal-warnings"
  = note: error: undefined symbol: __gxx_personality_v0 (referenced by top-level compiled C/C++ code)
          warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
          warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
          warning: ___gxx_personality_v0 may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
          Error: Aborting compilation due to previous errors
          emcc: error: 'C:/Users/camer/Documents/Tech/emsdk/node/14.15.5_64bit/bin/node.exe C:\Users\camer\Documents\Tech\emsdk\upstream\emscripten\src\compiler.js C:\Users\camer\AppData\Local\Temp\tmp6dw5_wby.txt' failed (returned 1)


error: aborting due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
Unwinding on `wasm32-unknown-emscripten` seems to be broken again, so this PR disables it. Undoes 62c3443e9659be2bf0ae53df1421c846ceaca904, fixes #85821, redux of #89762 which was abandoned by its author.

### CODE REVIEW

**Comment by nagisa** [comment]
Given that this is not the first time this default changes, I think we should have a write-up detailing a final decision and stick with it, instead of switching it back and forth every time the upstream project releases a new version.

**Review by vincenzopalazzo — APPROVED** [review]
LGTM, but lets wait the final decision on that problem

**Comment by Jules-Bertholet** [comment]
From my limited sleuthing:
According to https://github.com/rust-lang/rust/issues/85821#issuecomment-851739742, apparently unwiding has never really worked properly on Emscripten. Emscripten offers [two alternative exception handing methods](https://emscripten.org/docs/porting/exceptions.html): via JS and natively via WASM. The JS method is slow, requires JS (which not all WASM environments have), and it's not clear how easy it would be to support with Rust. The WASM exceptions seem much more straightforward, but they rely on a not-yet-finalized proposal that most WASM engines don't support.

**Comment by nagisa** [comment]
I'm going to nominate this for a T-compiler discussion.

I think it is a correct thing to disable unwinding here, but I worry that doing so will break somebody's CI or something. Probably is worth it over a target that doesn't work at all, but also if we make a decision like this now, I feel like we should make it such that it is permanent. That is, we would commit to `wasm32-unknown-emscripten` always using the abort panic scheme. We'd commit here so that users of this target can actually rely on a specific panicking mechanism here rather than being forced to specify `--panic=abort` to get any semblance of stability here.

**Comment by Jules-Bertholet** [comment]
https://github.com/rust-lang/rust/blob/2fa9789f596dd7639e1a242d466637b53179f4d1/compiler/rustc_target/src/spec/wasm_base.rs#L83-L87

I don't think committing to never ever having any form of unwinding for the Emscripten target makes sense, given that Rust doesn't commit to that for the other Wasm targets; once the [Wasm exception handling proposal](https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md) is finalized and [stabilized everywhere](https://webassembly.org/roadmap/), it can be used to implement unwinding. A weaker commitment which I think does make sense is that unwinding won't be implemented with any mechanism other than native Wasm exceptions, as Emscripten's homegrown exception handling doesn't offer any sort of stable API that Rust can rely on.

**Comment by Jules-Bertholet** [comment]
Another option is to gate WebAssembly unwinding behind an `exceptions` target feature, or some similar mechanism. See also https://github.com/rust-lang/rust/issues/77839

**Comment by apiraino** [comment]
The topic of disabling unwind has been discussed during T-compiler meeting on Zulip ([notes](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-04-21.html#279690372)). One idea was to disable unwind (so this PR can move forward) and define a process involving a writeup to reinstate it back if/when deemed necessary.

@rustbot label -I-compiler-nominated

File: compiler/rustc_codegen_ssa/src/back/linker.rs [hunk 1]
```diff
@@ -1132,11 +1132,13 @@ impl<'a> Linker for EmLinker<'a> {
...
         // Preserve names or generate source maps depending on debug info
-        self.cmd.arg(match self.sess.opts.debuginfo {
-            DebugInfo::None => "-g0",
-            DebugInfo::Limited => "-g3",
-            DebugInfo::Full => "-g4",
-        });
+        let args: &[&str] = match self.sess.opts.debuginfo {
+            DebugInfo::None => &["-g0"],
+            DebugInfo::Limited => &["-g3"],
+            DebugInfo::Full => &["-g3", "-gsource-map"],
```
Comment (nagisa) [inline compiler/rustc_codegen_ssa/src/back/linker.rs:1138]: This change should ideally have been a separate PR.

File: src/tools/compiletest/src/header.rs [hunk 2]
```diff
@@ -860,7 +861,7 @@ pub fn make_test_description<R: Read>(
         ignore |= !has_tsan && config.parse_name_directive(ln, "needs-sanitizer-thread");
         ignore |= !has_hwasan && config.parse_name_directive(ln, "needs-sanitizer-hwaddress");
         ignore |= !has_memtag && config.parse_name_directive(ln, "needs-sanitizer-memtag");
-        ignore |= config.target_panic == PanicStrategy::Abort
+        ignore |= (config.target_panic == PanicStrategy::Abort || !has_unwind_support)
```
Comment (nagisa) [inline src/tools/compiletest/src/header.rs:864]: Why is the `config.target_panic` check insufficient here? My intuition is that the targets in question should output `panic="abort"` for `rustc --print=cfg`.

File: library/core/tests/num/wrapping.rs [hunk 3]
```diff
@@ -64,14 +64,12 @@ wrapping_test!(test_wrapping_i8, i8, i8::MIN, i8::MAX);
 wrapping_test!(test_wrapping_i16, i16, i16::MIN, i16::MAX);
 wrapping_test!(test_wrapping_i32, i32, i32::MIN, i32::MAX);
 wrapping_test!(test_wrapping_i64, i64, i64::MIN, i64::MAX);
-#[cfg(not(target_os = "emscripten"))]
```
Comment (nagisa) [inline library/core/tests/num/wrapping.rs:67]: I feel like this should've been split out into a separate PR, since i128 support has nothing to do with unwinding.

File: library/alloc/benches/btree/map.rs [hunk 4]
```diff
@@ -349,11 +349,13 @@ pub fn iter_100(b: &mut Bencher) {
 }
 
 #[bench]
+#[cfg_attr(target_arch = "wasm32", ignore)] // not enough memory
```
Comment (nagisa) [inline library/alloc/benches/btree/map.rs:352]: Hm, this seems quite arbitrary. Doesn't WASM have `memory.grow` which should be invoked by the allocator to allocate memory pages past the initial allocation?

I feel like this ignore is ignoring a potentially genuine problem or that this test fails to work on 32-bit targets in general and should be rewritten to work on these targets (or ignored if pointer size is 32-bit.)

File: library/alloc/tests/vec.rs [hunk 5]
```diff
@@ -293,6 +293,7 @@ fn test_retain() {
 }
 
 #[test]
+#[cfg_attr(not(panic = "unwind"), should_panic)]
```
Comment (nagisa) [inline library/alloc/tests/vec.rs:296]: Hm, this is an interesting way to annotate these. Are traps actually caught on wasm like this?

File: src/bootstrap/native.rs [hunk 6]
```diff
@@ -736,10 +736,6 @@ impl Step for TestHelpers {
         builder.info("Building test helpers");
         t!(fs::create_dir_all(&dst));
         let mut cfg = cc::Build::new();
-        // FIXME: Workaround for https://github.com/emscripten-core/emscripten/issues/9013
```
Comment (nagisa) [inline src/bootstrap/native.rs:739]: This seems like it ought to be its own PR.

File: src/ci/docker/scripts/emscripten.sh [hunk 7]
```diff
@@ -20,5 +20,5 @@ exit 1
 
 git clone https://github.com/emscripten-core/emsdk.git /emsdk-portable
 cd /emsdk-portable
-hide_output ./emsdk install 1.39.20
-./emsdk activate 1.39.20
+hide_output ./emsdk install 2.0.17
```
Comment (nagisa) [inline src/ci/docker/scripts/emscripten.sh:23]: Similarly this seems like it ought to be its own PR. I'm not qualified to comment on what version of emscripten we want to target our support towards, but general rule of thumb has been so far to use a low version to ensure that we support a wide variety of versions and don't accidentally depend on very new features.

File: library/alloc/benches/btree/map.rs (hunk 4 above)
Comment (Jules-Bertholet) [reply library/alloc/benches/btree/map.rs:352]: By default, Emscripten doesn't use `memory.grow`; it needs to be [enabled via a flag](https://emscripten.org/docs/optimizing/Optimizing-Code.html#memory-growth). Maybe Rust should enable that flag by default?

Comment (nagisa) [reply library/alloc/benches/btree/map.rs:352]: Sounds to me like it might be a good idea, but also sounds like out of scope for this specific PR.

File: src/tools/compiletest/src/header.rs (hunk 2 above)
Comment (Jules-Bertholet) [reply src/tools/compiletest/src/header.rs:871]: `config.target_panic` apparently doesn't get set based off of `rustc --print=cfg`, only base off of command-line flags.

File: library/alloc/tests/vec.rs (hunk 5 above)
Comment (Jules-Bertholet) [reply library/alloc/tests/vec.rs:296]: This just tells the test harness to ignore the test on non-unwind targets, because `should_panic` tests are ignores when `panic != "unwind"`. `#[cfg_attr(not(panic = "unwind"), ignore)]` would work just as well

File: src/tools/compiletest/src/header.rs (hunk 2 above)
Comment (nagisa) [reply src/tools/compiletest/src/header.rs:871]: Got it. We should still be trying to use `--print=cfg` output here, but I guess it works for now.

**Comment by tlively** [comment]
Hi, I work on Emscripten and have historically worked on the Emscripten beckends in Rust, too. I am the one who previously enabled unwinding for the Emscripten target. From briefly looking at the issues that motivated this, it looks like this reporters were trying to use versions of Emscripten that use versions of LLVM that are incompatible with Rust's version of LLVM. In other words, I think the root problem here is a lack of documentation or automation about what version of Emscripten should be used with Rust, not that there is an actual bug with Emscripten or Rust.

**Comment by Jules-Bertholet** [comment]
No, this wasn't an LLVM compatibility issue as far as I can tell, this was due to a change in Emscripten's `libunwind`.

**Comment by tlively** [comment]
Sure, but the principle is the same. Using Emscripten versions from before that change to Emscripten's `libunwind` is the "correct" fix to this problem, since the latest version (or any other recent version) of Emscripten is not compatible with Rust in general.

**Comment by Jules-Bertholet** [comment]
I would think that the "correct" fix would be to rely only on APIs that the Emscripten project makes stability guarantees for, so users don't need to be bound to a specific version.

**Comment by tlively** [comment]
I agree that that would be even better, but Emscripten does not currently make stability guarantees for anything beyond the basic C ABI, so unfortunately that is not realistic right now. We're aiming to stabilize our C++ ABI soon, which would include libunwind, but it is not stable yet. Relatedly, we would like to stabilize a dynamic linking ABI soon, but again, that's not stable yet.
